server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -252,8 +280,14 @@ public void onFailure(String source, Exception e) {
 
             @Override
             public void clusterStateProcessed(String source, ClusterState oldState, final ClusterState newState) {
-                logger.info("snapshot [{}] started", snapshot);
-                listener.onResponse(snapshot);
+                try {
+                    logger.info("snapshot [{}] started", snapshot);
+                    listener.onResponse(snapshot);
+                } finally {
+                    if (newEntry.state().completed() || newEntry.shards().isEmpty()) {
```

Comment:
No need to actually run the full check here over all entries in `newState` here, the only two options of ending the snapshot here right away are if it has no shards or was set to `FAILED` right away.